src/main/java/br/com/desstecnologia/desstecnologia/Dao/ClienteDao.java

```diff
@@ -0,0 +1,55 @@
+package br.com.desstecnologia.desstecnologia.Dao;
+
+import br.com.desstecnologia.desstecnologia.model.Clientes;
+
+import java.util.HashMap;
+
+public class ClienteDao {
+
+    private static ClienteDao instance;
+    private HashMap<Integer, Clientes> clientesList;
+
+    private Integer key = 1;
+
+    private ClienteDao() { }
+
+    public void save(Clientes cliente) {
+        clientesListInstance();
+        clientesList.put(key, cliente);
+        key++;
+    }
+
+    public void delete(Integer id) {
+        clientesListInstance();
+        clientesList.remove(id);
+    }
+
+    public void merge(Integer id, Clientes cliente) {
+        clientesListInstance();
+        clientesList.put(id, cliente);
+    }
+
+    public Clientes findById(Integer id) {
+        clientesListInstance();
+        return clientesList.get(id);
+    }
+
+    public HashMap<Integer, Clientes> findAll() {
+        clientesListInstance();
+        return clientesList;
```

Comment:
Uma discussão interessante aqui! Retornar o `HashMap` privado direto aqui expõe ele para quem chamar. Imagina o seguinte, você tem um método que usa esse singleton e faz uma coisa tipo assim:

```java
ClienteDao dao = ClienteDao.getInstance();
HashMap<Integer, Clientes> clientes = dao.findAll();
clientes.put(id, new Cliente());
```

Como o `HashMap` é retornado como referência, esse código vai alterar os dados do singleton sem usar os métodos que deveriam ser utilizados, tipo `save()`.

Uma forma de evitar esse problema é retornar uma cópia do `HashMap`, tipo assim:

```java
public HashMap<Integer, Clientes> findAll() {
    return new HashMap<Integer, Clientes>(clientesList);
}
```

Então agora se o mapa for alterado, essa mudança será apenas local e não vai afetar o singleton.

Aí a próxima discussão é a diferença de cópia rasa e profunda (shallow copy e deep copy). Geralmente as cópias são rasas, como no exemplo que eu mostrei acima. Isso significa que o `HashMap` vai ser diferente, mas os objetos no mapa ainda são o mesmo, então alterar o cliente aí vai afetar o singleton.

Nesse caso o `HashMap` é só um atalho, em um projeto real teria um banco de verdade, mas é um ponto importante de lembrar.